### PR TITLE
Added bonus/feature by category for Skills

### PR DESCRIPTION
**Pull request by crnormand**
Added bonuses by category for Skills.   This one is kind of messy because of the necessity of passing the categories around.

1.  The "get*Bonus*" methods in GURPSCharacter now accept a category string (actually, a single string of all categories separated by commas), and when they are comparing the bonus to see if it matches, they now also check the category criteria.

2.  To shorten some of the criteria checking code, they can now respond directly if they are "an exact match" kind of criteria (Is), or an "anything goes" kind (isAnything).

3.  SkillBonus gained a category criteria and all of the internal methods were updated to include it (object creation, copy, XML read/write, etc.).   Also, a SkillBonus can now determine if it matchesCategories().

4.  The SkillBonusEditor gained a category criteria row.

5.  Skill was updated to include its category in all of its level calculation methods.

6.  And the SkillEditor was modified to pass in the category when determining level.

7.  And finally, Technique and the TechniqueEditor were modified to include the categories as well, since they subclass from Skill.

7a.   And I forgot a tooltip method on the TechniqueEditor, so it will show tooltips just like the SkillEditor.

I am sorry this one is hairier than the previous PRs.   I tried to whittle out as much as possible, but except for #2 and #7a (above), everything is necessary to make this work.

And unfortunately, the following PRs (for Spells and WeaponDamage) are going to be about the same.

### CODE REVIEW

File: src/com/trollworks/gcs/character/GURPSCharacter.java [hunk 1]
```diff
@@ -2724,7 +2724,7 @@ public int getIntegerBonusFor(String id, StringBuilder toolTip) {
      * @param specializationQualifier The specialization qualifier.
      * @return The bonus.
      */
-    public int getSkillComparedIntegerBonusFor(String id, String nameQualifier, String specializationQualifier) {
+    public int getSkillComparedIntegerBonusFor(String id, String nameQualifier, String specializationQualifier, String categoryQualifier) {
         return getSkillComparedIntegerBonusFor(id, nameQualifier, specializationQualifier, null);
```
Comment (richardwilkes) [inline src/com/trollworks/gcs/character/GURPSCharacter.java]: This just calls itself, in an infinite loop

File: src/com/trollworks/gcs/character/GURPSCharacter.java [hunk 2]
```diff
@@ -2724,7 +2724,7 @@ public int getIntegerBonusFor(String id, StringBuilder toolTip) {
      * @param specializationQualifier The specialization qualifier.
      * @return The bonus.
      */
-    public int getSkillComparedIntegerBonusFor(String id, String nameQualifier, String specializationQualifier) {
+    public int getSkillComparedIntegerBonusFor(String id, String nameQualifier, String specializationQualifier, String categoryQualifier) {
```
Comment (richardwilkes) [inline src/com/trollworks/gcs/character/GURPSCharacter.java]: Need to update the JavaDoc to include the new parameter

File: src/com/trollworks/gcs/character/GURPSCharacter.java [hunk 3]
```diff
@@ -2735,14 +2735,14 @@ public int getSkillComparedIntegerBonusFor(String id, String nameQualifier, Stri
      * @param toolTip                 The toolTip being built
      * @return The bonus.
      */
-    public int getSkillComparedIntegerBonusFor(String id, String nameQualifier, String specializationQualifier, StringBuilder toolTip) {
+    public int getSkillComparedIntegerBonusFor(String id, String nameQualifier, String specializationQualifier, String categoryQualifier, StringBuilder toolTip) {
```
Comment (richardwilkes) [inline src/com/trollworks/gcs/character/GURPSCharacter.java]: Javadoc again

File: src/com/trollworks/gcs/feature/SkillBonus.java [hunk 4]
```diff
@@ -98,28 +103,41 @@ public String getKey() {
         return buffer.toString();
     }
 
+    public boolean matchesCategories(String categories) {
+        String[] cats = categories.split(COMMA);
+        for (String category : cats) {
+            if (mCategoryCriteria.matches(category.trim())) {
```
Comment (richardwilkes) [inline src/com/trollworks/gcs/feature/SkillBonus.java]: Should probably do a case-insensitive comparison

File: src/com/trollworks/gcs/skill/Skill.java [hunk 5]
```diff
@@ -869,7 +869,7 @@ private SkillDefault getBestDefault(SkillDefault excludedDefault) {
                         int level = skillDefault.getType().getSkillLevel(character, skillDefault, excludes);
                         if (skillDefault.getType().isSkillBased()) {
                             String name = skillDefault.getName();
-                            level -= character.getSkillComparedIntegerBonusFor(ID_NAME + ASTERISK, name, skillDefault.getSpecialization());
+                            level -= character.getSkillComparedIntegerBonusFor(ID_NAME + ASTERISK, name, skillDefault.getSpecialization(), "");  //$NON-NLS-1$
```
Comment (richardwilkes) [inline src/com/trollworks/gcs/skill/Skill.java]: Doesn't this need to know the categories to do the right thing?

File: src/com/trollworks/gcs/feature/SkillBonus.java (hunk 4 above)
Comment (crnormand) [reply src/com/trollworks/gcs/feature/SkillBonus.java]: Actually, it is already case insensitive.   The StringCriteria code handles that.

File: src/com/trollworks/gcs/character/GURPSCharacter.java (hunk 1 above)
Comment (crnormand) [reply src/com/trollworks/gcs/character/GURPSCharacter.java]: Duh.... sorry, was trying to cull out just these changed from my other instance of Eclipse.   It has been fixed in the latest commit.

File: src/com/trollworks/gcs/skill/Skill.java (hunk 5 above)
Comment (crnormand) [reply src/com/trollworks/gcs/skill/Skill.java]: I was uncertain.    The problem is that SkillDefault did not have a category, so I assumed none.   But I see your point, on line 872, we are trying to find the level of the skill that we are basing this default on... which is different that the Skill from which we are trying to calculate the SkillDefault for.

I think I have it fixed.   It looks up the Skill for the SkillDefault and uses that Skill's categories when asking the character for the correct level for that skill..

Unfortunately, it doesn't correctly update the tooltip :-(    The default is calculated before the Skill and any modifiers that have affected the default are lost.   Not a huge problem (i can already see how to fix it), but I will leave that to the next PR.

**Comment by crnormand** [comment]
I kept the categories as a string because of the editors.  While in the editors, the category is just a string, so I would have to encode it into the set of categories.   To me it seemed 6 of one, 1/2 dozen of the other.

I do agree that if we keep the "categories as string" parameter, that I should consolidate all of the category composition/decomposition/matching/etc. login in the same class.        Which is what I have done with this latest commit.

I would like to keep it this way (as a String vs. Set<String>) with your permission.   All of the code for composing and decomposing the categories is kept in ListRow, now.  If you feel strongly against, I can refactor and have the editors call the composition method to make the Set<String> instead.

Concerning the two category matching methods, they actually support slightly different functionality.   hasCategory() is solely for the TextTemplate output (which has the funky logic of being able to match on the category name, or the name up to ":").   The other (now implemented in ListRow.matchesCategories()) is used by the criteria system, which could include any kind of "is/is not/begins with", etc.    

I could change the hasCategory() method to accept a StringCriteria insead, but I wasn't sure we would want to include the funky logic check with our generic criteria matching.   It was specifically designed for a FantasyGrounds feature where are can split advantages and equipment by category.   When we want to split "coins" from the rest of the inventory, there was a problem because many of the coins in the data libraries have the category "Money: GURPS Standard Medieval Coins", but the DFRPG library has the category set to "Money".   So being able to match on "Money" was the only way, with the data as is, catch as many of the coin instances as possible.

Hence why there are 2 different "matching" methods.

File: src/com/trollworks/gcs/character/GURPSCharacter.java [hunk 6]
```diff
@@ -2722,17 +2722,19 @@ public int getIntegerBonusFor(String id, StringBuilder toolTip) {
      * @param id                      The feature ID to search for.
      * @param nameQualifier           The name qualifier.
      * @param specializationQualifier The specialization qualifier.
+     * @parem categoryQualifier The categories qualifier
```
Comment (richardwilkes) [inline src/com/trollworks/gcs/character/GURPSCharacter.java]: `@parem` -> `@param` ... surprised Eclipse didn't complain about this

File: src/com/trollworks/gcs/character/GURPSCharacter.java [hunk 7]
```diff
@@ -2722,17 +2722,19 @@ public int getIntegerBonusFor(String id, StringBuilder toolTip) {
...
-        return getSkillComparedIntegerBonusFor(id, nameQualifier, specializationQualifier, null);
+        return getSkillComparedIntegerBonusFor(id, nameQualifier, specializationQualifier, categoryQualifier, null);
     }
 
     /**
      * @param id                      The feature ID to search for.
      * @param nameQualifier           The name qualifier.
      * @param specializationQualifier The specialization qualifier.
-     * @param toolTip                 The toolTip being built
+     * @parem categoryQualifier The categories qualifier
```
Comment (richardwilkes) [inline src/com/trollworks/gcs/character/GURPSCharacter.java]: Same here

**Comment by crnormand** [comment]
Understood, and I will get on it this evening.

File: src/com/trollworks/gcs/character/GURPSCharacter.java (hunk 7 above)
Comment (crnormand) [reply src/com/trollworks/gcs/character/GURPSCharacter.java]: Ok, I think I have spell checking enabled correctly... although that is no guarantee that I will spell things right.   ;-)

Comment (richardwilkes) [reply src/com/trollworks/gcs/character/GURPSCharacter.java]: The JavaDoc verifier should actually be throwing a fit on these... but I must admit I haven't specifically tested that theory out.

**Comment by crnormand** [comment]
Done.   The latest commit passes the categories as Set<String> instead of string.

**Pull request merged**